examples/color_palette/src/main.rs

```diff
@@ -0,0 +1,643 @@
+                        )
+                        .width(Length::Units(150))
+                        .size(14)
+                        .padding(2),
+                    ),
+            )
+            .push(
+                Row::new()
+                    .spacing(10)
+                    .push(Text::new("Lab").width(Length::Units(50)))
+                    .push(Slider::new(lab1, 0.0..=100.0, lab.l, move |l| {
+                        Message::LabColorChanged(palette::Lab { l, ..lab })
+                    }))
+                    .push(Slider::new(lab2, -128.0..=127.0, lab.a, move |a| {
+                        Message::LabColorChanged(palette::Lab { a, ..lab })
+                    }))
+                    .push(Slider::new(lab3, -128.0..=127.0, lab.b, move |b| {
+                        Message::LabColorChanged(palette::Lab { b, ..lab })
+                    }))
+                    .push(
+                        TextInput::new(
+                            &mut self.lab_text_state,
+                            "",
+                            &mut self.lab_text_value,
+                            |_s| Message::TextInput,
+                        )
+                        .width(Length::Units(150))
+                        .size(14)
+                        .padding(2),
+                    ),
+            )
+            .push(
+                Row::new()
+                    .spacing(10)
+                    .push(Text::new("Lch").width(Length::Units(50)))
+                    .push(Slider::new(lch1, 0.0..=100.0, lch.l, move |l| {
+                        Message::LchColorChanged(palette::Lch { l, ..lch })
+                    }))
+                    .push(Slider::new(
+                        lch2,
+                        0.0..=128.0,
+                        lch.chroma,
+                        move |chroma| {
+                            Message::LchColorChanged(palette::Lch {
+                                chroma,
+                                ..lch
+                            })
+                        },
+                    ))
+                    .push(Slider::new(
+                        lch3,
+                        0.0..=360.0,
+                        lch.hue.to_positive_degrees(),
+                        move |hue| {
+                            Message::LchColorChanged(palette::Lch {
+                                hue: palette::LabHue::from_degrees(hue),
+                                ..lch
+                            })
+                        },
+                    ))
+                    .push(
+                        TextInput::new(
+                            &mut self.lch_text_state,
+                            "",
+                            &mut self.lch_text_value,
+                            |_s| Message::TextInput,
+                        )
+                        .width(Length::Units(150))
+                        .size(14)
+                        .padding(2),
+                    ),
```

Comment:
I think we could get rid of duplication if we use a `ColorPicker` struct to contain the state of the 3 sliders and we use a custom trait to abstract the specifics of each color space:

```rust
struct ColorPicker<C: ColorSpace> {
    sliders: [slider::State; 3],
    color_space: PhantomData<C>,
}

trait ColorSpace: Sized {   
    const COMPONENT_RANGES: [RangeInclusive<f32>; 3];

    fn new(a: f32, b: f32, c: f32) -> Self;

    fn components(&self) -> (f32, f32, f32);

    fn to_string(&self) -> String;
}
```

I believe this should allow us to implement `ColorPicker::view`, like so:

```rust
impl<C: ColorSpace> ColorPicker<C> {
    fn view(&mut self, color: &C) -> Element<C> {
        let (a, b, c) = color.components();
        // ...
    }
}
```

Then, in `Application::view` we could simply:

```rust
Column::new()
    // ...
    .push(self.rgb.view(&srgb).map(Message::RgbColorChanged))
    .push(self.hsl.view(&hsl).map(Message::HslColorChanged))
    // ...
```

I know this may be a bit involved! No pressure. I can try to give it a shot myself.

Reply:
Good idea! I've made a bit of progress toward this design on my local copy. Need to figure out a couple things.

Reply:
Converted into something like this in 430f78a. Total lines of code actually increased 😆 , but we don't have the repetitive view code any longer.